openedx/course-discovery
# feat: add missing pieces for Course data modified timestamp workflow

### [PROD-3408](https://2u-internal.atlassian.net/browse/PROD-3408)

### Description
This PR is a followup work from the PR https://github.com/openedx/course-discovery/pull/3976. This PR adds the necessary elements to complete course data modified timestamp update flow.

#### Seat Model
Seat model is linked with CourseRun and contains information about seat pricing and upgrade deadline override (this field is available on publisher in CourseRunForm section). Appropriate field tracker  based signal handler is added to update course data modified timestamp when seat related attrs are changed.

#### Course url slug
Course url slug and its history are maintained in CourseUrlSlug model. The slug handling also differs if the course is draft or non-draft. The draft->non-draft changes are not managed via set_official_state util. Instead, set_active_url_slug method is responsible for appropriate slug and history updates. Therefore,  in the method,  the mechanism to update course data modified timestamp has been added in the method set_active_url_slug. **That method did not have any unit tests, so I had to add them as well to verify the timestamp updates are working as expected.**

#### M2M relationships 

Following two categories of M2M relationships were not covered before:
- SortedM2M Field
- Taggable Manager

The details of the M2M relationship complexity can be seen in the internal doc https://2u-internal.atlassian.net/wiki/spaces/CAT/pages/356156060/Timestamp+Based+Data+Query. However, I am adding some concise context here.

Rest framework’s serializer [update() method use M2M field’s .set()](https://github.com/encode/django-rest-framework/blob/master/rest_framework/serializers.py#L1010) method to add M2M relationship after updating the instance object The set() method in [M2MField in Django](https://github.com/django/django/blob/main/django/db/models/fields/related_descriptors.py#L895) varies the m2m setting behavior based on clear flag. However, for SortedM2M, the clear is[ always set to True](https://github.com/jazzband/django-sortedm2m/blob/e646aafa81416c7888a7d358c7293bd4f707a25f/sortedm2m/fields.py#L47-L50), to maintain the ordering of the objects in M2M relationship.  **The workaround here is to get M2M relationship pks from non-draft versions of Course/CourseRun and then compare those Pks with values being added in m2m_relationship.**

Taggable manager in django-taggit uses [TaggedItem](https://github.com/jazzband/django-taggit/blob/master/taggit/models.py#L185) to store information on which tags are assigned to which Django model (via content type). TaggedItem is the [default](https://github.com/jazzband/django-taggit/blob/master/taggit/managers.py#L440) through value if it is not defined explicitly. In Discovery course_metadata app, the value for through is not defined whenever TaggableManager is used (be it Org, Course, ProductMeta, Program, etc). Now, to receive m2m_changed signal, we need to specify the sender. The sender is i[ntermediate or through model ](https://docs.djangoproject.com/en/4.2/ref/signals/#m2m-changed)(which is TaggedItem for tag manager). So, therefore, when m2m_changed is attempted to be set only for Course.topics.through, since through=TaggedItem, whenever m2m for tags is changed on any other model, say ProductMeta, the m2m_changed handler is called because through is the same for Course & ProductMeta. **Explicit model label checks helped curate the proper handler for each taggable manager**

#### Product Value & Location Restriction
In Course api, the handling for product value & location restriction is updated to rely on calling save() instead of .filter().update() to enforce pre_save signal handler execution.

### Testing

- Make sure you have one ExecEd course in your local discovery for complete flow testinng
- Check out this branch.
- Run the latest publisher on your local. 
- Go to any course. 
- Edit its url slug. Ensure the slug is changed and the timestamp is updated.  Verify for both draft & non-draft versions
- Edit the course collaborators (both add/remove) and verify the timestamp updates
- Edit the course subjects (both add/remove) and verify the timestamp updates
- Edit product value and verify data timestamp updates
- Add/Edit location restriction and verify data timestamp updates
- In course run form, update verified upgrade deadline (only for Verified types) and verify the course timestamp is updated
- Edit the course run staff (both add/remove) and verify the timestamp updates
- For last step in testing, edit any field in Course/CourseRun form and ensure the data timestamp is updated. If any field change does not do that, flag that during review.
